pkg/worker/worker.go

```diff
@@ -0,0 +1,177 @@
+func NewWorker(config conf.Config) Worker {
+	return &worker{
+		logger: log.NewLogger(config),
+		jobs:   make(map[string]*Job),
+	}
+}
+
+// worker implementation.
+type worker struct {
+	// logger is responsible to handle the
+	// stdout and stderr of a running process
+	logger log.Logger
+	// jobs is concurrency safe map to store
+	// the requested jobs
+	jobs map[string]*Job
+	// mtx to control jobs concurret access
+	mtx sync.RWMutex
+}
+
+// Start runs a Linux single command with arguments.
+// If the command runs with success, a Job identifier will be returned.
+// A log file will be created with the Job id name to capture the stdout and stderr
+// of a running process.
+// To get the process status, the Job request will be stored in memory,
+// and a goroutine will be launched to update the job status when the process is finished.
+func (w *worker) Start(command Command) (jobID string, err error) {
+	cmd := exec.Command(command.Name, command.Args...)
+	jobID = uuid.NewString()
+	logfile, err := w.logger.Create(jobID)
+	if err != nil {
+		return
+	}
+	// redirect the stdout and stderr to the log file
+	cmd.Stdout = logfile
+	cmd.Stderr = logfile
+	if err = cmd.Start(); err != nil {
+		w.logger.Remove(jobID)
+		return
+	}
+	// create and store the job
+	w.mtx.Lock()
+	w.jobs[jobID] = &Job{ID: jobID, Cmd: cmd, Status: &Status{Pid: cmd.Process.Pid}}
+	w.mtx.Unlock()
+	// update the job status in background after
+	// the process exit
+	go w.updateJob(jobID)
+	return
+}
+
+// Stop terminates a running Job gracefully sending a SIGTERM to the process.
+// If the job doesn't exitis an error will be returned.
+func (w *worker) Stop(jobID string) (err error) {
+	w.mtx.RLock()
+	defer w.mtx.RUnlock()
+	job, ok := w.jobs[jobID]
+	if !ok {
+		return errors.New("Job not found")
+	}
+	return job.Cmd.Process.Signal(syscall.SIGTERM)
+}
+
+// Query returns the process status of a specific Job.
+// If the job doesn't exitis an error will be returned.
+func (w *worker) Query(jobID string) (status Status, err error) {
+	w.mtx.RLock()
+	defer w.mtx.RUnlock()
+	job, ok := w.jobs[jobID]
+	if !ok {
+		return Status{}, errors.New("Job not found")
+	}
+	status = *job.Status
+	return
+}
+
+// Stream reads from the log file, like 'tail -f' through
+// a channel. If the context is canceled the channel will
+// be closed and the tailing will be stopped.
+func (w *worker) Stream(ctx context.Context, jobID string) (logchan chan string, err error) {
+	w.mtx.RLock()
+	job, ok := w.jobs[jobID]
+	w.mtx.RUnlock()
+	if !ok {
+		return nil, errors.New("Job not found")
+	}
+	return w.logger.Tailf(ctx, job.ID)
+}
+
+// updateJob access the Job information stored in memory and
+// waits for the command to exit to get the process status.
+// the log file will be removed from the file system after
+// process is finished.
+func (w *worker) updateJob(jobID string) {
+	w.mtx.RLock()
+	job, ok := w.jobs[jobID]
+	w.mtx.RUnlock()
+	if !ok {
+		return
+	}
```

Comment:
You could just pass the job object from Start here since you already have it, then this wouldn't be needed.